pinot-controller/src/main/java/org/apache/pinot/controller/workload/scheme/PropagationUtils.java

```diff
@@ -33,37 +33,72 @@
 import org.apache.pinot.spi.config.table.TableConfig;
 import org.apache.pinot.spi.config.table.TableType;
 import org.apache.pinot.spi.config.table.TenantConfig;
+import org.apache.pinot.spi.config.workload.InstanceCost;
 import org.apache.pinot.spi.config.workload.NodeConfig;
+import org.apache.pinot.spi.config.workload.PropagationEntity;
 import org.apache.pinot.spi.config.workload.PropagationScheme;
 import org.apache.pinot.spi.config.workload.QueryWorkloadConfig;
 import org.apache.pinot.spi.utils.builder.TableNameBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 
 /**
- * This class provides utility methods for workload propagation.
+ * Provides utility methods for workload propagation in Pinot.
+ *
+ * <p>
+ * This class centralizes logic for resolving Helix tags, instance mappings, and matching query
+ * workload configs to their applicable propagation scope.
+ * </p>
  */
 public class PropagationUtils {
-
+  private static final Logger LOGGER = LoggerFactory.getLogger(PropagationUtils.class);
   private PropagationUtils() {
   }
 
   /**
-   * Get the mapping tableNameWithType → {BROKER_NODE→brokerTag, SERVER_NODE→(serverTag + overrides)}
-   * 1. Get all table configs from the PinotHelixResourceManager
-   * 2. For each table config, extract the tenant config
-   * 3. For each tenant config, get the broker and server tags
-   * 4. Populate the helix tags for BROKER_NODE and SERVER_NODE separately
+   * Builds a mapping from table name with type to Helix tags per node type.
+   *
+   * <p>For each table:</p>
+   * <ol>
+   *   <li>Fetch the {@link TenantConfig}.</li>
+   *   <li>Derive broker and server tags from the tenant configuration.</li>
+   *   <li>Associate {@link NodeConfig.Type#BROKER_NODE} with broker tags.</li>
+   *   <li>Associate {@link NodeConfig.Type#SERVER_NODE} with server tags (consuming and/or completed
+   *       for realtime, or offline for batch).</li>
+   * </ol>
+   *
+   * @param pinotResourceManager Resource manager used to fetch table configs.
+   * @return A mapping of tableNameWithType to node type → Helix tags.
    */
   public static Map<String, Map<NodeConfig.Type, Set<String>>> getTableToHelixTags(
           PinotHelixResourceManager pinotResourceManager) {
     Map<String, Map<NodeConfig.Type, Set<String>>> tableToTags = new HashMap<>();
-    for (TableConfig tableConfig : pinotResourceManager.getAllTableConfigs()) {
+    List<TableConfig> tableConfigs = pinotResourceManager.getAllTableConfigs();
+    if (tableConfigs == null) {
```

Comment:
getAllTableConfigs never returns a null. It returns an emptyList. 

Can we check all such error handling places and clean up the code a little bit?